File: haskell/glue/src/Glue/Lib/Bool/Ge.hs
module Glue.Lib.Bool.Ge where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

ge :: IR Eval
ge = NativeFunc geImpl

geImpl :: IR Eval -> Eval (IR Eval)
geImpl a = pure $ NativeFunc (geRight a)

geRight :: IR Eval -> IR Eval -> Eval (IR Eval)
geRight a b = do
    va <- eval a
    vb <- eval b
    case (va, vb) of
        (Integer na, Integer nb) -> pure . Bool $ na >= nb
        (Float na, Float nb) -> pure . Bool $ na >= nb
        (Integer na, Float nb) -> pure . Bool $ fromIntegral na >= nb
        (Float na, Integer nb) -> pure . Bool $ na >= fromIntegral nb
        _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/bool/ge.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Greater than or equal comparison function
/// Mirrors Haskell Glue.Lib.Bool.Ge.ge exactly
final Ir ge = IrNativeFunc(geImpl);

/// Greater than or equal comparison implementation
/// Mirrors Haskell Glue.Lib.Bool.Ge.geImpl exactly
Eval<Ir> geImpl(Ir a) {
  return Eval.pure(IrNativeFunc(geRight(a)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Bool.Ge.geRight exactly
Eval<Ir> Function(Ir) geRight(Ir a) {
  return (Ir b) => switch ((a, b)) {
    (IrInteger(value: final na), IrInteger(value: final nb)) => Eval.pure(
      IrBool(na >= nb),
    ),
    (IrFloat(value: final na), IrFloat(value: final nb)) => Eval.pure(
      IrBool(na >= nb),
    ),
    (IrInteger(value: final na), IrFloat(value: final nb)) => Eval.pure(
      IrBool(na >= nb),
    ),
    (IrFloat(value: final na), IrInteger(value: final nb)) => Eval.pure(
      IrBool(na >= nb),
    ),
    _ => throwError(wrongArgumentType(['number', 'number'])),
  };
}
